homeassistant/components/drop_connect/config_flow.py

```diff
@@ -81,14 +79,9 @@ async def async_step_mqtt(self, discovery_info: MqttServiceInfo) -> FlowResult:
         self.__hub_id = topicElements[2]
         self.__device_id = topicElements[3]
 
-        # Build the unique ID and register it.
-        self.__unique_id = f"{self.__hub_id}_{self.__device_id}"
-        await self.async_set_unique_id(self.__unique_id)
-
+        await self.async_set_unique_id(f"{self.__hub_id}_{self.__device_id}")
         # Abort if this device has already been configured.
-        already_configured = self._async_current_ids(False)
-        if self.__unique_id in already_configured:
-            _LOGGER.debug("Device %s already configured", self.__unique_id)
+        if self.unique_id in self._async_current_ids(False):
             return self.async_abort(reason="invalid_discovery_info")
```

Comment:
```suggestion
        self._abort_if_unique_id_configured()
```

Not sure if I understand what you are trying to achieve, I am not familiar with `_async_current_ids`, but wouldn't the following method work for this scenario?

Reply:
see https://github.com/home-assistant/core/pull/104319#discussion_r1411266730

Reply:
The DROP hub will publish its discovery data via MQTT any time the dataset changes. (e.g. the user adds or removes a device from the hub) If I use `self._abort_if_unique_id_configured()` and the hub publishes discovery data, the config flow fires again (triggered via MQTT discovery) and I end up with duplicate devices in the HA UI. The approach I'm using tests to see if the device in question is already configured, not just if it was already handled by this particular config flow run.

I don't know enough to assert that my approach is correct, but I do know that using `self._abort_if_unique_id_configured()` creates a mess of duplicate devices.

Reply:
the `_abort_if_unique_id_configured()` does exactly what you are describing - it checks if any config entry exists with same `self.unique_id` (_that's why first call `async_set_unique_id()` to have it set_), to do so it also calls `_async_current_entries(include_ignore=True)` (_you might recognize the different `include_ignore` option_).

Another alternative is to check the result of `async_set_unique_id()` because it "Returns optionally existing config entry with same ID" - example:

```python
    existing_entry = await self.async_set_unique_id(f"{self.__hub_id}_{self.__device_id}")
    if existing_entry is not None:
        return self.async_abort(reason="already_configured")
```

we should also use `already_configured` instead of `invalid_discovery_info`

Reply:
Why do we not use the MQTT discovery features instead?

Reply:
https://www.home-assistant.io/integrations/sensor.mqtt/

and 

https://www.home-assistant.io/integrations/mqtt/#how-to-use-discovery-messages.

I am happy to help to get working examples.

It saves a lot of energy. Branding can be done too.

Reply:
I forgot to mention that removing devices is supported out of the box.

Reply:
Part of our goal is to keep the MQTT API on the DROP hub simple and platform-agnostic.  There are other automation platforms we eventually want to target and implementing an API for each isn't practical given the constraints in the embedded firmware.  I also suspect I may not even have enough free stack space on the microcontroller to generate the necessary discovery messages.  I'm not denying that there are benefits to the approach you are recommending, but I've been working with reviewers for two weeks refining this PR into the state that it is now in.  I was under the impression that it was very close to acceptance and my firmware development is done and proven.  Unless changing the fundamental approach I am using is required, I would like to hold this course.

Reply:
I went ahead and built a solution in the DROP hub firmware that uses the HA MQTT integration discovery feature to "integrate" DROP with Home Assistant.  I did this primarily so that I could evaluate the option fairly.  At this point, I could provide integration with Home Assistant in our product line either with or without an actual HA integration.  Here is a quick pro/con analysis for each approach from my perspective:

### Using the custom integration in this pull request:
**PRO:**
 - DROP product branding, documentation in HA, and visibility within HA's ecosystem.
 - MQTT API in the DROP firmware is simpler and less verbose.
 - MQTT API in the DROP firmware can be re-used for other automation platforms.
 - Potential future functionality is not limited by the MQTT discovery mechanism.

**CON:**
 - Significant added maintenance time in the form of pull requests in the HA repos and requirements added by HA reviewers, such as needing to create and publish a Python library just to parse some JSON data.
 
### Using the MQTT integration discovery:
**PRO:**
 - No need for a custom integration, which means releasing new DROP products or API endpoints is much easier.

**CON:**
 - No product branding or other typical visibility in HA.  DROP products are hidden under the MQTT integration.
 - No custom functionality like dynamic icons that change based on switch state.
 - MQTT API in the DROP firmware is HA-specific and more verbose.
 - Device linking defined using 'via_device' does not work after initial discovery until HA is restarted.  (potential bug?)
 - Added complexity in the DROP firmware that is otherwise unnecessary.
 - Feels like an unprofessional shortcut and creates a less polished user experience.
 - May limit functionality we may want to add in the future.
 
It appears that the only compelling reason to use the existing MQTT discovery is ease of maintenance from our end, but even that feels short-sighted. (Admittedly, it is also very tempting.)  I don't think the argument that it avoids code duplication is valid because the integration provides significant advantages that the existing MQTT discovery does not.  At this point, we will likely only take the native MQTT discovery route if we cannot get this PR approved in a timely fashion.
 
My impression is that the discovery feature in the MQTT integration is an excellent tool for small projects or one-off custom integrations, but for a commercial product line like DROP it leaves a lot to be desired in terms of customizability and user experience.

Reply:
Based on the result of the discussion at [home-assistant/architecture#1006](https://github.com/home-assistant/architecture/discussions/1006) and an evaluation of my options, I'm marking this resolved.  A custom integration is the best option for the DROP product line.